### PR TITLE
Advanced config

### PR DESCRIPTION
**Pull request by DvogelHallowelt**

### CODE REVIEW

**Review by osnard — CHANGES_REQUESTED** [review]

File: src/Command/Analyze.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+				$analyzer->setInput( $this->input );
+			}
+
+			$gerritUserQuestion = new Question( 'Question test: ', '' );
+			$gerritUserQuestion->setValidator( function ( $answer ) {
+			$answer = trim( $answer );
+			if ( empty( $answer ) ) { throw  new \RuntimeException( "Required!" );
+			}
+			return $answer;
+			} );
```
[the remark this answers is not in the document]
Comment (DvogelHallowelt) [reply src/Command/Analyze.php:53]: I had to stash my work. Of course this had already been removed.

File: src/Analyzer/ConfluenceAnalyzer.php [hunk 2]
```diff
@@ -10,16 +10,20 @@
...
 use Psr\Log\LoggerInterface;
 use Psr\Log\NullLogger;
 use SplFileInfo;
+use Symfony\Component\Console\Helper\QuestionHelper;
+use Symfony\Component\Console\Input\Input;
 use Symfony\Component\Console\Output\Output;
+use Symfony\Component\Console\Question\Question;
 
-class ConfluenceAnalyzer extends AnalyzerBase implements LoggerAwareInterface, IOutputAwareInterface {
+class ConfluenceAnalyzer extends AnalyzerBase implements LoggerAwareInterface, IOutputAwareInterface, IUserInteraction {
```
[the remark this answers is not in the document]
Comment (DvogelHallowelt) [reply src/Analyzer/ConfluenceAnalyzer.php:26]: No, your right. Import script is asking questions now.

File: src/IUserInteraction.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace HalloWelt\MigrateConfluence;
+
+use Symfony\Component\Console\Helper\QuestionHelper;
+use Symfony\Component\Console\Input\Input;
+use Symfony\Component\Console\Output\Output;
+
+interface IUserInteraction {
```
[the remark this answers is not in the document]
Comment (DvogelHallowelt) [reply src/IUserInteraction.php:9]: Okay, I will move it

File: src/Analyzer/ConfluenceAnalyzer.php [hunk 4]
```diff
@@ -59,6 +68,31 @@ class ConfluenceAnalyzer extends AnalyzerBase implements LoggerAwareInterface, I
...
+
+	/**
+	 * @var array
+	 */
+	private $advancedConfig = [];
+
+	/**
+	 * @var bool
+	 */
+	private $hasAdvancedConfig = true;
```
Comment (osnard) [inline src/Analyzer/ConfluenceAnalyzer.php:94]: Shouldn't we assume that there is _no_ advanced config by default?

File: src/Analyzer/ConfluenceAnalyzer.php [hunk 5]
```diff
@@ -112,8 +161,14 @@ public function analyze( SplFileInfo $file ): bool {
 			return true;
 		}
 
+		if ( isset( $this->config['config'] ) ) {
+			$this->advancedConfig = $this->config['config'];
+			$this->hasAdvancedConfig = true;
```
Comment (osnard) [inline src/Analyzer/ConfluenceAnalyzer.php:166]: ... especially as there does not seem to be code to set it to `false`...

File: src/Analyzer/ConfluenceAnalyzer.php [hunk 6]
```diff
@@ -128,6 +183,51 @@ protected function doAnalyze( SplFileInfo $file ): bool {
 		$this->dom->load( $file->getPathname() );
 		$this->helper = new XMLHelper( $this->dom );
 
+		if ( !$this->hasAdvancedConfig ) {
```
Comment (osnard) [inline src/Analyzer/ConfluenceAnalyzer.php:186]: So we require the advanced config to contain _all_ required information, right? No partial config can be passed, making the tool ask for additional input?

That's fine, but we need to document this and maybe provide a `config.sample.yml`.

File: src/Command/Analyze.php [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+				);
+			}
+			if ( $analyzer instanceof IOutputAwareInterface ) {
+				$analyzer->setOutput( $this->output );
+			}
+			if ( $analyzer instanceof IUserInteraction ) {
+				$analyzer->setQuestionHelper( $this->getHelper( 'question' ) );
+				$analyzer->setOutput( $this->output );
+				$analyzer->setInput( $this->input );
+			}
```
Comment (osnard) [inline src/Command/Analyze.php:70]: I still believe this should be done by the base class. This subclass should just do something like 

```php
$this->readConfigFile( $this->config );
return parent::doProcessFile();
```

File: src/Command/Convert.php [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+		);
+		$this->targetPathname = preg_replace( '#\.mraw$#', '.wiki', $this->targetPathname );
+	}
+
+	private function ensureTargetPath() {
+		$baseTargetPath = dirname( $this->targetPathname );
+		if ( !file_exists( $baseTargetPath ) ) {
+			mkdir( $baseTargetPath, 0755, true );
+		}
+	}
```
Comment (osnard) [inline src/Command/Convert.php:87]: Too much code copied from the base class

File: src/Converter/ConvertableEntities/Image.php [hunk 9]
```diff
@@ -157,6 +164,15 @@ private function makeImageLinkWithDebugInfo( DOMDocument $dom, array $params, $c
...
 			$debug = " ###BROKENIMAGE $confluenceFileKey ###";
+		} elseif ( $this->nsFileRepoCompat === true ) {
+			$filename = $params[0];
+			$pos = strpos( $filename, '_' );
+			if ( $pos !== false ) {
+				$namespace = substr( $filename, 0, $pos );
+				if ( $namespace !== false ) {
+					$params[0] = str_replace( $namespace . '_', $namespace . ':', $filename );
+				}
+			}
```
Comment (osnard) [inline src/Converter/ConvertableEntities/Image.php:175]: Can I please see an UnitTest for this?

File: src/Extractor/ConfluenceExtractor.php [hunk 10]
```diff
@@ -30,6 +35,10 @@ protected function doExtract( SplFileInfo $file ): bool {
 		$this->dom->load( $file->getPathname() );
 		$this->helper = new XMLHelper( $this->dom );
 
+		if ( isset( $this->config['config']['categories'] ) ) {
+			$this->categories = $this->config['config']['categories'];
+		}
```
Comment (osnard) [inline src/Extractor/ConfluenceExtractor.php:40]: Would it make sense to allow different categories for different spaces?

Not necessary though...

**Review by osnard — CHANGES_REQUESTED** [review]

File: src/Analyzer/ConfluenceAnalyzer.php (hunk 4 above)
Comment (DvogelHallowelt) [reply src/Analyzer/ConfluenceAnalyzer.php]: Set to false by default

File: src/Analyzer/ConfluenceAnalyzer.php (hunk 5 above)
Comment (DvogelHallowelt) [reply src/Analyzer/ConfluenceAnalyzer.php:166]: done

File: src/Analyzer/ConfluenceAnalyzer.php (hunk 6 above)
Comment (DvogelHallowelt) [reply src/Analyzer/ConfluenceAnalyzer.php:186]: Yes, that's right

File: src/Command/Convert.php (hunk 8 above)
Comment (DvogelHallowelt) [reply src/Command/Convert.php:46]: done

File: src/Converter/ConvertableEntities/Image.php (hunk 9 above)
Comment (DvogelHallowelt) [reply src/Converter/ConvertableEntities/Image.php:175]: Sure

File: src/Extractor/ConfluenceExtractor.php (hunk 10 above)
Comment (DvogelHallowelt) [reply src/Extractor/ConfluenceExtractor.php:40]: No, not now

**Pull request merged**